modules/userId.js

```diff
@@ -0,0 +1,431 @@
+ * @property {Object} value - all object properties will be appended to the User ID bid data
+ */
+
+/**
+ * @typedef {Object} SubmoduleStorage
+ * @property {string} type - browser storage type (html5 or cookie)
+ * @property {string} name - key name to use when saving/reading to local storage or cookies
+ * @property {number} expires - time to live for browser cookie
+ */
+
+/**
+ * @typedef {Object} SubmoduleParams
+ * @property {string} partner - partner url param value
+ * @property {string} url - webservice request url used to load Id data
+ */
+
+/**
+ * @typedef {Object} Submodule
+ * @property {string} name - submodule and config have matching name prop
+ * @property {decode} decode - decode a stored value for passing to bid requests
+ * @property {getId} getId - performs action to obtain id and return a value in the callback's response argument
+ */
+
+/**
+ * @callback getId
+ * @param {SubmoduleParams} [submoduleConfigParams]
+ * @param {Object} [consentData]
+ * @returns {(Function|Object|string)}
+ */
+
+/**
+ * @callback decode
+ * @param {Object|string} idData
+ * @returns {Object}
+ */
+
+/**
+ * @typedef {Object} SubmoduleContainer
+ * @property {Submodule} submodule
+ * @property {SubmoduleConfig} submoduleConfig
+ * @property {Object} idObj - decoded User ID data that will be appended to bids
+ * @property {function} callback
+ */
+
+const MODULE_NAME = 'User ID';
+const COOKIE = 'cookie';
+const LOCAL_STORAGE = 'html5';
+const DEFAULT_SYNC_DELAY = 500;
+
+// @type {number} delay after auction to make webrequests for id data
+export let syncDelay;
+
+// @type {SubmoduleContainer[]}
+export let submodules;
+
+// @type {SubmoduleContainer[]}
+let initializedSubmodules;
+
+// @type {Submodule}
+export const unifiedIdSubmodule = {
+  name: 'unifiedId',
+  decode(value) {
+    return {
+      'tdid': value['TDID']
+    }
+  },
+  getId(submoduleConfigParams, consentData) {
+    const partner = (submoduleConfigParams && typeof submoduleConfigParams.partner === 'string') ? submoduleConfigParams.partner : 'prebid';
+    const url = (submoduleConfigParams && typeof submoduleConfigParams.url === 'string') ? submoduleConfigParams.url : `http://match.adsrvr.org/track/rid?ttd_pid=${partner}&fmt=json`;
+
+    return function (callback) {
+      ajax(url, response => {
+        let responseObj;
+        if (response) {
+          try {
+            responseObj = JSON.parse(response);
+          } catch (error) {
+            utils.logError(error);
+          }
+        }
+        callback(responseObj);
+      }, undefined, { method: 'GET' });
+    }
+  }
+};
+
+// @type {Submodule}
+export const pubCommonIdSubmodule = {
+  name: 'pubCommonId',
+  decode(value) {
+    return {
+      'pubcid': value
+    }
+  },
+  getId() {
+    // If the page includes its own pubcid object, then use that instead.
+    if (typeof window['PublisherCommonId'] === 'object') {
+      return window['PublisherCommonId'].getId();
```

Comment:
should wrap this in a `try/catch` since we don't know for sure if the function exists or if it throws and error.